File: android/libraries/rib-compiler-test/src/test/java/com/uber/rib/compiler/InteractorTestGeneratorProcessorTestBase.java
/*
 * Copyright (C) 2017. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.compiler;

import androidx.annotation.NonNull;

import com.google.common.collect.ImmutableSet;
import com.google.testing.compile.JavaFileObjects;

import org.junit.Before;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import javax.annotation.processing.ProcessingEnvironment;
import javax.tools.JavaFileObject;

public abstract class InteractorTestGeneratorProcessorTestBase {

  private RibProcessor ribProcessor;
  private ArrayList<JavaFileObject> sources;

  @Before
  public void setup() {
    ribProcessor =
        new RibProcessor() {
          InteractorTestGenerator interactorTestGenerator;

          @Override
          public synchronized void init(ProcessingEnvironment processingEnv) {
            interactorTestGenerator = new InteractorTestGenerator(processingEnv, errorReporter);
            super.init(processingEnv);
          }

          @NonNull
          @Override
          protected List<ProcessorPipeline> getProcessorPipelines(
              @NonNull ProcessContext processContext) {
            return Collections.<ProcessorPipeline>singletonList(
                new RibInteractorProcessorPipeline(processContext, interactorTestGenerator));
          }

          @Override
          public Set<String> getSupportedAnnotationTypes() {
            return ImmutableSet.of(
                RibInteractorProcessorPipeline.SUPPORT_ANNOTATION_TYPE.getCanonicalName());
          }
        };
    sources = new ArrayList<>();
  }

  protected RibProcessor getRibInteractorProcessor() {
    return ribProcessor;
  }

  protected ArrayList<JavaFileObject> getSources() {
    return sources;
  }

  protected void addResourceToSources(@NonNull String file) {
    getSources().add(getResourceFile(file));
  }

  @NonNull
  protected JavaFileObject getResourceFile(@NonNull String file) {
    return JavaFileObjects.forResource(file);
  }
}

File: android/libraries/rib-compiler-test/src/test/kotlin/com/uber/rib/compiler/InteractorTestGeneratorProcessorTestBase.kt
/*
 * Copyright (C) 2017. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.compiler

import com.google.common.collect.ImmutableSet
import com.google.testing.compile.JavaFileObjects
import java.util.ArrayList
import javax.annotation.processing.ProcessingEnvironment
import javax.tools.JavaFileObject
import org.junit.Before

abstract class InteractorTestGeneratorProcessorTestBase {
  protected lateinit var ribInteractorProcessor: RibProcessor
  protected lateinit var sources: ArrayList<JavaFileObject>

  @Before
  fun setup() {
    ribInteractorProcessor =
      object : RibProcessor() {
        var interactorTestGenerator: InteractorTestGenerator? = null

        @Synchronized
        override fun init(processingEnv: ProcessingEnvironment) {
          interactorTestGenerator =
            InteractorTestGenerator(processingEnv, ErrorReporter(processingEnv.messager))
          super.init(processingEnv)
        }

        override fun getProcessorPipelines(
          processContext: ProcessContext,
        ): List<ProcessorPipeline> {
          return listOf(
            RibInteractorProcessorPipeline(processContext, interactorTestGenerator),
          )
        }

        override fun getSupportedAnnotationTypes(): Set<String> {
          return ImmutableSet.of(
            RibInteractorProcessorPipeline.SUPPORT_ANNOTATION_TYPE.canonicalName,
          )
        }
      }
    sources = ArrayList()
  }

  protected fun addResourceToSources(file: String) {
    sources.add(getResourceFile(file))
  }

  protected fun getResourceFile(file: String): JavaFileObject {
    return JavaFileObjects.forResource(file)
  }
}
